supermarq-benchmarks/supermarq/qcvv/su2.py

```diff
@@ -0,0 +1,269 @@
+        """
+        records = []
+        for sample in samples:
+            records.append(
+                {
+                    "num_two_qubit_gates": sample.data["num_two_qubit_gates"],
+                    **sample.probabilities,
+                }
+            )
+
+        return pd.DataFrame(records)
+
+    def analyze_results(self, plot_results: bool = True) -> SU2Results:
+        """Perform the experiment analysis and store the results in the `results` attribute.
+
+        Args:
+            plot_results: Whether to generate plots of the results. Defaults to False.
+
+        Returns:
+            A named tuple of the final results from the experiment.
+        """
+        fit = linregress(
+            x=self.raw_data["num_two_qubit_gates"],
+            y=np.log(self.raw_data["00"] - 1 / 4),
+            # Scale the y coordinate to account for limit of the decay being 1/4
+        )
+        gate_fid = np.exp(fit.slope)
+        gate_fid_std = fit.stderr * gate_fid
+
+        single_qubit_noise = 1 - 4 / 3 * np.exp(fit.intercept)
+        single_qubit_noise_std = fit.intercept_stderr * (1 - single_qubit_noise)
+
+        self._results = SU2Results(
+            target="& ".join(self.targets),
+            total_circuits=len(self.samples),
+            two_qubit_gate_fidelity=gate_fid,
+            two_qubit_gate_fidelity_std=gate_fid_std,
+            single_qubit_noise=single_qubit_noise,
+            single_qubit_noise_std=single_qubit_noise_std,
+        )
+
+        if plot_results:
+            self.plot_results()
+
+        return self.results
+
+    @staticmethod
+    def _haar_random_rotation() -> cirq.Gate:
+        """Returns:
+        Haar randomly sampled SU(2) rotation.
+        """
+        gate: cirq.Gate | None = None
+        while gate is None:
+            gate = cirq.single_qubit_matrix_to_phxz(cirq.testing.random_special_unitary(dim=2))
+        return gate
+
+    def _component(self, include_two_qubit_gate: bool) -> cirq.Circuit:
+        """Core component of the experimental circuits.
+
+        These circuits that are repeated to create the full circuit. Can optionally include the
+        two qubit gate being measured, as is required for the
+        first half of the full circuit, but not for the second half.
+
+        The component looks like:
+        .. code::
+
+            0: ───R1───Q───X───Q───
+                       │       │
+            1: ───R2───Q───X───Q───
+
+        where :code:`R1` and :code:`R2` are Haar randomly chosen SU(2) rotation
+        and :code:`Q-Q` represents the two qubit gate being measured.
+
+        Args:
+            include_two_qubit_gate: Whether to include the two qubit gate being measured
+
+        Returns:
+            The sub circuit to be repeated when building the full circuit
+        """
+        return cirq.Circuit(
+            self._haar_random_rotation().on(self.qubits[0]),
+            self._haar_random_rotation().on(self.qubits[1]),
```

Comment:
i believe the same rotation is supposed to be applied to both gates, so that they can be done with a single global pulse on systems that support it (the language in the paper is "random global single-qubit rotations between sequences of CZ entangling gates")

Reply:
Sorry only just seen this comment.
Yes I agree, thanks. I must have misinterpreted the paper. Now fixed 😃